File: 1319-number-of-operations-to-make-network-connected.ts
import UnionFind from "./union-find"

function makeConnected(n: number, connections: number[][]): number {
  if (connections.length < n - 1) {
    return -1
  }
  const uf = new UnionFind(n)
  for (const [a, b] of connections) {
    uf.union(a, b)
  }
  const disjointSets = new Set()
  for (let i = 0; i < n; i++) {
    const root = uf.findRoot(i)
    disjointSets.add(root)
  }
  return disjointSets.size - 1
}

File: problems/1319_number_of_operations_to_make_network_connected.py
from utils.union_find import UnionFind


class Solution:
    def makeConnected(self, n: int, connections: list[list[int]]) -> int:
        if len(connections) < n - 1:
            return -1
        uf = UnionFind(n)
        for [a, b] in connections:
            uf.union(a, b)
        distinct_sets: set[int] = set()
        for i in range(n):
            root = uf.find_root(i)
            distinct_sets.add(root)
        return len(distinct_sets) - 1
